### PR TITLE
Add support for partial sync

### PR DESCRIPTION
**Pull request by austinzheng**
TODO:
- [x] Swift support
- [x] Rebase this onto something a little more sensible

### CODE REVIEW

File: Realm/RLMRealm+Sync.mm [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#import "results.hpp"
+#import "sync/partial_sync.hpp"
+#import "shared_realm.hpp"
+
+using namespace realm;
+
+@implementation RLMRealm (Sync)
+
+- (void)fetchResultsForQuery:(NSString *)query objectType:(Class)type callback:(RLMPartialSyncFetchCallback)callback {
+    NSString *className = NSStringFromClass(type);
```
Comment (austinzheng) [inline Realm/RLMRealm+Sync.mm]: Needs to check if the Realm is actually a partially synced Realm, and throw/return early if not.

Comment (bdash) [reply Realm/RLMRealm+Sync.mm]: I'd suggest pushing that checking down to object store if possible.

File: Realm/RLMRealm+Sync.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+@class RLMResults;
+
+/**
+ A callback used to vend the results of a partial sync fetch.
+ */
+typedef void(^RLMPartialSyncFetchCallback)(RLMResults * _Nullable results, NSError * _Nullable error);
+
+NS_ASSUME_NONNULL_BEGIN
+
+///
```
Comment (austinzheng) [inline Realm/RLMRealm+Sync.h]: Should this be blank, or a `:nodoc:`?

Comment (bdash) [reply Realm/RLMRealm+Sync.h]: Blank. `:nodoc:` hides everything in the category from the documentation, which isn't what we want.

**Comment by bdash** [comment]
CI has been having issues this evening, and I _think_ that's all the object server build issues are (they're timing out before the tests even run).

**Pull request merged**